app/grandchallenge/algorithms/tasks.py

```diff
@@ -163,105 +159,137 @@ def create_algorithm_jobs(
         The algorithm image to use for new job
     algorithm_model
         The algorithm model to use for the new job or None
 
     Returns
     -------
-    Filtered set of ComponentInterfaceValues
+    Dictionary of valid ArchiveItems for new jobs, grouped by AlgorithmInterface
     """
     from grandchallenge.algorithms.models import Job
+    from grandchallenge.evaluation.models import (
+        get_archive_items_for_interfaces,
+        get_valid_jobs_for_interfaces_and_archive_items,
+    )
 
-    input_interfaces = {*algorithm_image.algorithm.inputs.all()}
+    if not archive_items:
+        return {}
 
-    existing_jobs = {
-        frozenset(j.inputs.all())
-        for j in Job.objects.filter(
-            algorithm_image=algorithm_image, algorithm_model=algorithm_model
+    archives = {ai.archive for ai in archive_items}
+    if len(archives) != 1:
+        raise RuntimeError(
+            "The archive items should all belong to the same archive."
         )
-        .annotate(
-            inputs_match_count=Count(
-                "inputs",
-                filter=Q(
-                    inputs__in={civ for civ_set in civ_sets for civ in civ_set}
-                ),
-            )
-        )
-        .filter(inputs_match_count=len(input_interfaces), creator=None)
-        .prefetch_related("inputs")
-    }
-
-    valid_job_inputs = []
-
-    for civ_set in civ_sets:
-        # Check interfaces are complete
-        civ_interfaces = {civ.interface for civ in civ_set}
-        if input_interfaces.issubset(civ_interfaces):
-            # If the algorithm works with a subset of the interfaces
-            # present in the set then only feed these through to the algorithm
-            valid_input = {
-                civ for civ in civ_set if civ.interface in input_interfaces
-            }
-        else:
-            continue
+    (archive,) = archives
 
-        # Check job has not been run
-        if frozenset(valid_input) in existing_jobs:
-            continue
+    algorithm_interfaces = (
+        algorithm_image.algorithm.interfaces.prefetch_related("inputs").all()
+    )
 
-        valid_job_inputs.append(valid_input)
+    # First, sort archive items by algorithm interface:
+    # An archive item is only valid for an interface if it has values
+    # for all inputs of the interface
+    valid_job_inputs = get_archive_items_for_interfaces(
+        algorithm_interfaces=algorithm_interfaces, archive_items=archive_items
+    )
+
+    # Next, get all system jobs that have been run for the provided archive items
+    # with the same model and image
+    if algorithm_model:
+        extra_filter = {"algorithm_model": algorithm_model}
+    else:
+        extra_filter = {"algorithm_model__isnull": True}
+
+    relevant_jobs = Job.objects.filter(
+        algorithm_image=algorithm_image,
+        algorithm_interface__in=valid_job_inputs.keys(),
+        creator=None,
+        inputs__archive_items__archive=archive,
```

Comment:
You have the archive items here, so you might be able to do:

```suggestion
        inputs__archive_items__in=archive_items,
```

Then, you might not need the `archive` from above.

Reply:
Ah that's a good point. Then indeed I don't need the archive above.